orte/mca/rml/ofi/rml_ofi_component.c

```diff
@@ -0,0 +1,1091 @@
+                    opal_output_verbose(1, orte_rml_base_framework.framework_output,
+                         "%s Received message on conduitid %d",
+                         ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
+                         conduit->conduit_id);
+                        // call the receive message handler that will call the rml_base
+                        ret = orte_rml_ofi_recv_handler(&wc, conduit->conduit_id);
+            } else if ( wc.flags & FI_MULTI_RECV ) {
+                    opal_output_verbose(1, orte_rml_base_framework.framework_output,
+                         "%s Received buffer overrun message on conduitid %d - need to repost",
+                         ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
+                         conduit->conduit_id);
+                        // reposting buffer
+                        ret = fi_recv(orte_rml_ofi.ofi_conduits[conduit->conduit_id].ep,
+                          orte_rml_ofi.ofi_conduits[conduit->conduit_id].rxbuf,
+                          orte_rml_ofi.ofi_conduits[conduit->conduit_id].rxbuf_size,
+                          fi_mr_desc(orte_rml_ofi.ofi_conduits[conduit->conduit_id].mr_multi_recv),
+                          0,&(conduit->rx_ctx1));  
+            }else {
+                opal_output_verbose(1,orte_rml_base_framework.framework_output,
+                        "CQ has unhandled completion event with FLAG wc.flags = 0x%x",
+                         wc.flags);
+            }
+        } else if (ret == -FI_EAVAIL) {
+            /**
+            * An error occured and is being reported via the CQ.
+            * Read the error and forward it to the upper layer.
+            */
+            opal_output_verbose(1, orte_rml_base_framework.framework_output,
+                         "%s cq_read for conduitid %d  returned error 0x%x <%s>",
+                         ORTE_NAME_PRINT(ORTE_PROC_MY_NAME),
+                         conduit->conduit_id, ret, 
+                         fi_strerror((int) -ret) );
+            ret = fi_cq_readerr(conduit->cq,
+                                &error,
+                                0);
+            if (0 > ret) {
+                opal_output_verbose(1,orte_rml_base_framework.framework_output,
+                                "Error returned from fi_cq_readerr: %zd", ret);
+                abort();
+            }
+            assert(error.op_context);
+            /* get the context from wc and call the error handler */
+            ofi_req = TO_OFI_REQ(error.op_context);
+            assert(ofi_req);
+            ret = orte_rml_ofi_error_callback(&error, ofi_req);
+            if (ORTE_SUCCESS != ret) {
+               opal_output_verbose(1,orte_rml_base_framework.framework_output,
+               "Error returned by request error callback: %zd",
+               ret);
+               abort();
+            }
+        } else {
+            /**
+             * The CQ is empty. Return.
+             */
```

Comment:
This is not quite correct -- according to https://ofiwg.github.io/libfabric/v1.3.0/man/fi_cq.3.html, the CQ is empty when `-FI_EAGAIN` is returned.  The logic here has you falling into this block for any negative value that is not `-FI_EAVAIL`, but I think that libfabric is allowed to return other error codes here, too.

Specifically:

* This block should only be entered when `-FI_EAGAIN` is returned
* There should be another block for handling other error codes returned from `fi_cq_read()`